File: Algorithm/Leetcode/Code_Materials/python_source/charpter16_slide/level2/length_of_longest_substring_two_distinct.py
# coding=utf-8
from collections import defaultdict

class LengthOfLongestSubstringTwoDistinct:
    def lengthOfLongestSubstringTwoDistinct(self, s):
        n = len(s)
        if n < 3:
            return n
        left, right = 0, 0
        hashmap = defaultdict()
        max_len = 2
        while right < n:
            if len(hashmap) < 3:
                hashmap[s[right]] = right
                right += 1
            if len(hashmap) == 3:
                del_idx = min(hashmap.values())
                del hashmap[s[del_idx]]
                left = del_idx + 1
            max_len = max(max_len, right - left)
        return max_len

if __name__ == '__main__':
    s = "eceba"
    lengthOfLongestSubstringTwoDistinct = LengthOfLongestSubstringTwoDistinct()
    print lengthOfLongestSubstringTwoDistinct.lengthOfLongestSubstringTwoDistinct(s)

File: Algorithm/Leetcode/Code_Materials/Java/src/main/java/com/yugutou/charpter16_slide/level2/LengthOfLongestSubstringTwoDistinct.java
package com.yugutou.charpter16_slide.level2;

import java.util.Collections;
import java.util.HashMap;

public class LengthOfLongestSubstringTwoDistinct {

    public static void main(String[] args) {
        String s = "eceba";
        System.out.println(lengthOfLongestSubstringTwoDistinct(s));
    }

    public static int lengthOfLongestSubstringTwoDistinct(String s) {

        if (s.length() < 3) {
            return s.length();
        }
        int left = 0, right = 0;
        HashMap<Character, Integer> hashmap = new HashMap<>();
        int maxLen = 2;

        while (right < s.length()) {

            if (hashmap.size() < 3)
                hashmap.put(s.charAt(right), right++);

            // 如果大小达到了3个
            if (hashmap.size() == 3) {
                // 最左侧要删除的位置
                int del_idx = Collections.min(hashmap.values());
                hashmap.remove(s.charAt(del_idx));
                // 窗口left的新位置
                left = del_idx + 1;
            }

            maxLen = Math.max(maxLen, right - left);
        }
        return maxLen;
    }
}
